src/main/java/com/scalyr/integrations/kafka/Event.java

```diff
@@ -169,18 +180,49 @@ public int estimatedSerializedBytes() {
       return estimatedSizeBytes;
     }
 
-    int size = getMessage() == null ? 0 : getMessage().length();
+    int size = Strings.isNullOrEmpty(getMessage()) ? 0 : estimateEscapedStringSize(getMessage());
     size += getTopic().length();
     size += EVENT_SERIALIZATION_OVERHEAD_BYTES;
 
     if (getAdditionalAttrs() != null) {
       size += getAdditionalAttrs().entrySet().stream()
-        .mapToInt(entry -> entry.getKey().length() + (entry.getValue() == null ? 0 : entry.getValue().toString().length())).sum();
+        .mapToInt(entry -> entry.getKey().length() + (entry.getValue() == null ? 0 : estimateEscapedStringSize(entry.getValue().toString()))).sum();
     }
     estimatedSizeBytes = size;
     return estimatedSizeBytes;
   }
 
+  /**
+   * Estimate the escaped String length for JSON strings.
+   * JSON payload requires extra escaping which needs to be taken into the size estimate.
+   *
+   * If it is not JSON, then use the String length as the estimate.
+   * This trades off some inaccuracies where the String may contain quotable characters for faster performance.
+   * @return Estimated escaped String size
+   */
+  private int estimateEscapedStringSize(String s) {
+    if (s == null) {
+      return 0;
+    }
+    if (s.length() < 2) {
+      return s.length();
+    }
+
+    if (s.charAt(0) == '{' && s.charAt(s.length() - 1) == '}') {
+      StringBuilder sb = jsonStringBuilder.get();
+      if (sb == null) {
+        sb = new StringBuilder(s.length() + 20);
+        jsonStringBuilder.compareAndSet(null, sb);
+      }
+      sb.setLength(0);
+
+      jsonStringEncoder.quoteAsString(s, sb);
```

Comment:
Do you know what the performance impact of this is?  For json logs, this seems like it's going double the amount of encoding work that needs to be done.. It's not clear to me what this does for the GC pressure as well.  Since it is going directly into a StringBuilder, maybe it's not so bad?  But, it does worry me.

Why not just use a simple hueristic in its place?  I suspect most of the extra characters come from escaping double quotes, newlines and carriage returns.  Why not just count those characters and get an estimate?

Reply:
Yeah my main concern also is a potential performance impact of this code.

Another question also is if and how this will work in real-life (e.g. the user could be sending us already pre-escaped JSON or similar (seen that before) and we wouldn't even go in this code path - although I guess if it's already pre escaped we probably wouldn't try to escape it again?), but I guess only way to check and test that is to get access to the raw payloads once we exceed the limit and audit them.

Perhaps we could gather some real life JSON payloads and run your implementation and the one Steven proposed on it and see if the estimates are close enough and if they are, use a simpler (and probably faster) version.

Reply:
Just realized in the middle of the night that re-using `StringBuilder` as a `static` will have multithreading issues.  I did this for performance to avoid allocating new Strings/StringBuffer.  We are single threaded in a single task (which is what I was thinking about yesterday), but in a single JVM we have multiple tasks on separate threads.  Will address this and performance questions in the morning.

Reply:
I think we should remove the the `StringBuilder` and `JsonStringEncoder`.  Possible alternatives are 1) Get a rough count estimate by string length + counting escapable characters. or 2) Don't try to estimate String escapable characters